File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        prev = None
        right = slow.next
        slow.next = None
        while right:
            temp = right.next
            right.next = prev
            prev = right
            right = temp
        
        left, right = head, prev
        while right:
            left_temp, right_temp = left.next, right.next
            left.next = right
            right.next = left_temp
            left, right = left_temp, right_temp

File: Data Structures & Algorithms/insertionSort/submission-0.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
    vector<vector<Pair>> insertionSort(vector<Pair>& pairs) {
        vector<vector<Pair>> res;
        for (int i = 0; i < pairs.size(); ++i)
        {
            int curr = i;
            int prev = curr - 1;
            while (prev >= 0 && pairs[curr].key < pairs[prev].key)
            {
                Pair temp = pairs[curr];
                pairs[curr] = pairs[prev];
                pairs[prev] = temp;
                --prev;
                --curr;
            }
            res.push_back(pairs);
        }
        return res;
    }
};
